### PR TITLE
Feature/profile

### PR DESCRIPTION
**Pull request by mlh-tiennd**
# Profile (example) #15

## What are you adding?
- [x] Store user in context
- [x] Display some user info in Profile screen
- [x] Form to update user info

## Breaking changes?
No

## Related PR
No

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -5,6 +5,8 @@
   "dependencies": {
     "@material-ui/core": "^4.11.0",
     "@material-ui/icons": "^4.9.1",
+    "@rematch/core": "^2.0.0-next.10",
+    "@rematch/loading": "^2.0.0-next.11",
```
Comment (yukinoda) [inline package.json:9]: @mlh-tiennd cc: @mlh-binhnt @themitvp 
Hi Tien, at the admin panel template we'll use Redux-Toolkit. So in our admin panel projects too.

Comment (mlh-tiennd) [reply package.json]: I finished converting from rematch to redux-toolkit. Please review this. Thanks!

**Review by danfirestone — APPROVED** [review]
LGTM

File: src/components/FormElements/FormikRadioGroup/FormikRadioGroup.tsx [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  const fieldIsInvalid = Object.prototype.hasOwnProperty.call(errors, name);
+
+  return (
+    <FormControl
+      fullWidth={fullWidth}
+      component="fieldset"
+      error={fieldIsInvalid}
+    >
+      <FormLabel component="legend">
+        {label || name.charAt(0).toLocaleUpperCase() + name.substr(1)}
```
Comment (themitvp) [inline src/components/FormElements/FormikRadioGroup/FormikRadioGroup.tsx:49]: use `??` in case the the label is an empty string, and do not manipulate the `name`, you are making assumptions about how it should look, but ideally we should just show it as we receive it.
```suggestion
        {label ?? name}
```

File: src/components/FormElements/FormikRadioGroup/FormikRadioGroup.tsx [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+import { RadioGroup } from "formik-material-ui";
+
+type RadioValue = string | number;
+
+type RadioOption =
+  | {
+      label: React.ReactNode;
+      value: RadioValue;
+    }
+  | RadioValue;
```
Comment (themitvp) [inline src/components/FormElements/FormikRadioGroup/FormikRadioGroup.tsx:19]: instead of having multiple options, it would be more simple to force it to always be a `label` and `value`, having these two different options makes it more complicated to maintain and provides very little benefit
```jsx
type RadioOption = {
  label: React.ReactNode;
  value: RadioValue;
}
```

File: src/components/FormElements/FormikSelect/FormikSelect.tsx [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+import { Select } from "formik-material-ui";
+
+type SelectOptionValue = string | number;
+
+type SelectOption =
+  | {
+      label: React.ReactNode;
+      value: SelectOptionValue;
+    }
+  | SelectOptionValue;
```
Comment (themitvp) [inline src/components/FormElements/FormikSelect/FormikSelect.tsx:18]: same here as for the radio group, only have the `label` and `value` option to keep it simple

File: src/components/FormElements/FormikSelect/FormikSelect.tsx [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  };
+
+  return (
+    <FormControl
+      component="fieldset"
+      error={fieldIsInvalid}
+      fullWidth={fullWidth}
+    >
+      <InputLabel htmlFor={name}>
+        {label || name.charAt(0).toLocaleUpperCase() + name.substr(1)}
```
Comment (themitvp) [inline src/components/FormElements/FormikSelect/FormikSelect.tsx:67]: same here as for the radio
```suggestion
        {label ?? name}
```

File: src/components/FormElements/FormikSwitch/FormikSwitch.tsx [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  name,
+  label,
+  ...rest
+}) => {
+  return (
+    <FormControlLabel
+      control={
+        <Field component={Switch} type="checkbox" name={name} {...rest} />
+      }
+      label={label || name.charAt(0).toLocaleUpperCase() + name.substr(1)}
```
Comment (themitvp) [inline src/components/FormElements/FormikSwitch/FormikSwitch.tsx:23]:
```suggestion
      label={label ?? name}
```

File: .eslintrc.js [hunk 7]
```diff
@@ -30,6 +30,7 @@ module.exports = {
     ],
     "react/jsx-props-no-spreading": 0,
     "import/prefer-default-export": 0,
+    "no-param-reassign": ["error", { props: false }],
```
Comment (themitvp) [inline .eslintrc.js:33]:
```suggestion
    // allow param reassign for redux-toolkit
    "no-param-reassign": ["error", { props: false }],
```

Comment (themitvp) [reply .eslintrc.js:33]: also you can add this:
```js
    // disabling circular dependency, as it is causing issues
    "import/no-cycle": 0,
```

File: .eslintrc.js [hunk 8]
```diff
@@ -38,7 +39,7 @@ module.exports = {
         patterns: [
           "@material-ui/*/*/*",
           "!@material-ui/core/test-utils/*",
-          "features/*/*/*",
+          // "features/*/*/*",
```
Comment (themitvp) [inline .eslintrc.js:42]: the pattern is there on purpose to make it clear that other features should not import stuff inside other features, only what is added to the barrel file should be used

Comment (mlh-tiennd) [reply .eslintrc.js]: Many thanks for reviewing! I finished.

**Comment by danfirestone** [comment]
@mlh-tiennd CC @themitvp @yukinoda 

I think we should be able to merge this in today. 

You guys happy to merge?

File: package.json [hunk 9]
```diff
@@ -86,6 +88,6 @@
     "lint-staged": "^10.4.0",
     "prettier": "^2.1.2",
     "react-app-rewired": "^2.1.6",
-    "typescript": "~3.7.2"
+    "typescript": "4.0.5"
```
Comment (themitvp) [inline package.json:91]: CRA v3 doesn't have support for Typescript v4.
In order to use TS v4, you need to use CRA v4, please undo this change

File: src/components/FormElements/FormikCheckbox/FormikCheckbox.tsx [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+import { CheckboxWithLabel } from "formik-material-ui";
+
+type CheckboxValue = string | number;
+
+type CheckboxOption =
+  | {
+      label: React.ReactNode;
+      value: CheckboxValue;
+    }
+  | CheckboxValue;
```
Comment (themitvp) [inline src/components/FormElements/FormikCheckbox/FormikCheckbox.tsx:18]: this should also only have `label` and `value` to simplify it

File: src/components/FormElements/FormikCheckbox/FormikCheckbox.tsx [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+  const fieldIsInvalid = Object.prototype.hasOwnProperty.call(errors, name);
+
+  return (
+    <FormControl
+      fullWidth={fullWidth}
+      component="fieldset"
+      error={fieldIsInvalid}
+    >
+      <FormLabel component="legend">
+        {label || name.charAt(0).toLocaleUpperCase() + name.substr(1)}
```
Comment (themitvp) [inline src/components/FormElements/FormikCheckbox/FormikCheckbox.tsx:46]:
```suggestion
        {label ?? name}
```

**Comment by yukinoda** [comment]
@mlh-tiennd Can you resolve the conflict?
@themitvp Can this PR be merged after the conflict is merged?

File: src/components/FormElements/FormikCheckbox/FormikCheckbox.tsx [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+          ) : (
+            <Field
+              component={CheckboxWithLabel}
+              {...rest}
+              name={name}
+              type="checkbox"
+              value={option}
+              Label={{ label: option }}
+            />
+          )
```
Comment (themitvp) [inline src/components/FormElements/FormikCheckbox/FormikCheckbox.tsx:64]: it will always be an object

**Review by themitvp — COMMENTED** [review]
@yukinoda after that is fixed then it can be merged

**Comment by mlh-tiennd** [comment]
> @mlh-tiennd Can you resolve the conflict?
> @themitvp Can this PR be merged after the conflict is merged?

**Comment by mlh-tiennd** [comment]
> @mlh-tiennd Can you resolve the conflict?
> @themitvp Can this PR be merged after the conflict is merged?

I resolved confict

**Review by yukinoda — APPROVED** [review]
LGTM

File: src/components/FormElements/FormikCheckbox/FormikCheckbox.tsx (hunk 12 above)
Comment (yukinoda) [reply src/components/FormElements/FormikCheckbox/FormikCheckbox.tsx]: cc: @mlh-tiennd @mlh-binhnt 
Resolved
https://github.com/monstar-lab-oss/admin-panel-template-reactjs/commit/df29289e3f14122c516b6541724c49f15664105c
